### PR TITLE
Refining Sunpath

### PR DESCRIPTION
**Pull request by devngc**
These changes have mainly been directed towards improving the altitude and azimuth calculation. @mostaphaRoudsari This is my first PR for a [+] version. Kindly let me know if I am supposed to provide any other file.

### CODE REVIEW

**Comment by mostaphaRoudsari** [comment]
@devngc thank you for the PR! This has been a long-time returning issue.

Do you have some test cases that compares the results of the code against NOAA solar calculator? I can also write the tests but I thought I should ask as you have probably already checked it during the process.

That are some PEP8 issues which I will comment them inside the code. Some of them are repetitive and all of them are easy to fix.

File: ladybug/sunpath.py [hunk 1]
```diff
@@ -151,36 +151,19 @@ def calculateSunFromDataTime(self, datetime, isSolarTime=False):
...
-        # RADIANS
-        zenith = math.acos(math.sin(self._latitude) * math.sin(solDec) +
-                           math.cos(self._latitude) * math.cos(solDec) *
-                           math.cos(math.radians(hourAngle)))
-
-        altitude = (math.pi / 2) - zenith
-
-        if hourAngle == 0.0 or hourAngle == -180.0 or hourAngle == 180.0:
+        # Degrees
+        zenith = math.degrees(math.acos(math.sin(self._latitude)*math.sin(math.radians(solDec))+math.cos(self._latitude)*math.cos(math.radians(solDec))*math.cos(math.radians(hourAngle))))
```
Comment (mostaphaRoudsari) [inline ladybug/sunpath.py]: 1. Add a white space before and after * - / +
2. Break down the line into smaller segments. We us 89 characters.

File: ladybug/sunpath.py [hunk 2]
```diff
@@ -151,36 +151,19 @@ def calculateSunFromDataTime(self, datetime, isSolarTime=False):
...
 
-            azimuth = math.pi if solDec < self._latitude else 0.0
+        altitude = 90 - zenith
 
+        # Degrees
+        if hourAngle>0:
```
Comment (mostaphaRoudsari) [inline ladybug/sunpath.py]: white space before and after >

File: ladybug/sunpath.py [hunk 3]
```diff
@@ -151,36 +151,19 @@ def calculateSunFromDataTime(self, datetime, isSolarTime=False):
...
+            azimuth = math.degrees(math.acos(((math.sin(self._latitude)*math.cos(math.raradians(zenith)))-math.sin(math.radians(solDec)))/(math.cos(self._l_latitude)*math.sin(math.radians(zenith)))))+180 % 360
```
Comment (mostaphaRoudsari) [inline ladybug/sunpath.py]: The same here for white-space and the length of the line.

File: ladybug/sunpath.py [hunk 4]
```diff
@@ -257,17 +240,75 @@ def _calculateSolarGeometry(self, datetime, year=2015):
...
         """
-        month, day, hour = datetime.month, datetime.day, datetime.floatHour
+        month, day, hour, minute = datetime.month, datetime.day, datetime.floatHour, datetime.minute
 
         a = 1 if (month < 3) else 0
         y = year + 4800 - a
         m = month + 12 * a - 3
 
-        julianDay = day + math.floor((153 * m + 2) / 5) + 59
+        def findFractionOf24(hour,minute):
```
Comment (mostaphaRoudsari) [inline ladybug/sunpath.py]: White space is missing after ,

File: ladybug/sunpath.py [hunk 5]
```diff
@@ -257,17 +240,75 @@ def _calculateSolarGeometry(self, datetime, year=2015):
...
+            """
+            This function calculates the fraction of the 24 hour the provided time represents
+            1440 is total the number of minutes in a 24 hour cycle.
+            args
```
Comment (mostaphaRoudsari) [inline ladybug/sunpath.py]: `:` is missing after args

File: ladybug/sunpath.py [hunk 6]
```diff
@@ -257,17 +240,75 @@ def _calculateSolarGeometry(self, datetime, year=2015):
...
+            This function calculates the number of days from 01-01-1900 to the provided date
+            args
+                year: Integer. The year in the date
+                month: Integer. The month in the date
+                day: Integer. The date
+            return: The number of days from 01-01-1900 to the date provided
+            """
+
+            # Making a list of years from the year 1900
+            years = [item for item in range(1900, year)]
```
Comment (mostaphaRoudsari) [inline ladybug/sunpath.py]: What about `years = range(1900, year)`

File: ladybug/sunpath.py [hunk 7]
```diff
@@ -257,17 +240,75 @@ def _calculateSolarGeometry(self, datetime, year=2015):
...
+
+            def is_leap_year(year):
```
Comment (mostaphaRoudsari) [inline ladybug/sunpath.py]: The name of the method should be def `isLeapYear`

**Comment by devngc** [comment]
Hi @mostaphaRoudsari,
Thank you for comments. I use Atom editor. Will look into these comments. Also, will try to find a package to do this PEP8 checking. 

Did you give it a try? Does it work at your end? and Does it produce more accurate altitude and azimuths?

**Comment by devngc** [comment]
@mostaphaRoudsari, 
I provided a link to GH file, which contains one test with NOAA results. I can document more cases in the code.

**Comment by mostaphaRoudsari** [comment]
Didn't check the GH file. Will do later today after my class. If you're using Atom then here are the packages that I use which includes autopep8:

**Comment by devngc** [comment]
@mostaphaRoudsari,
Thank you for sharing your list of packages. I installed them. Unfortunately, they crash at my end. I will keep looking for possible solutions.

File: ladybug/sunpath.py [hunk 8]
```diff
@@ -144,43 +144,57 @@ def calculateSunFromDataTime(self, datetime, isSolarTime=False):
...
-
-        altitude = (math.pi / 2) - zenith
+        if solTime / 4 < 0:
+            hourAngle = solTime / 4 + 180
+        else:
+            hourAngle = solTime / 4 - 180
 
-        if hourAngle == 0.0 or hourAngle == -180.0 or hourAngle == 180.0:
+        # Degrees
+        zenith = math.degrees(math.acos \
```
Comment (mostaphaRoudsari) [inline ladybug/sunpath.py]: This is not major but you don't need `\` inside parentheses.

File: ladybug/sunpath.py [hunk 9]
```diff
@@ -683,3 +756,21 @@ def __repr__(self):
             self.sunVector.y,
             self.sunVector.z
         )
+
+
+# Following are test cases to compare altitude and azimuth values with NOAA calculator
```
Comment (mostaphaRoudsari) [inline ladybug/sunpath.py]: Thanks for adding these cases. They either need to be written as a proper test cases or at least we should make sure that they won't be execute on import. To achieve that you can use `if __name__ == '__main__:`. If you don't know the concept watch [this video](https://www.youtube.com/watch?v=sugvnHA7ElY).

Once you add the if statement I can merge this. I will change them to proper test cases afterwards and you can follow that as an example for later.

Thanks again for all the great work @devngc!

**Comment by devngc** [comment]
I managed to run those packages eventually. So this file is checked for PEP8 for issues.

Now, speaking of test cases, the main() is not there in the file. Do you want me to create that function and put all the code inside it? I thought it may invite changes in other files that depend on this one so I thought I would ask first. Also, this will again raise line length related PEP8 issues.

For the time being, I commented out the test cases.

Please let me know if I should put everything in a main().

Also, In order to run this file to check everything, I need to write
`import __init__ as ladybug`
or else it throws import error. Just sharing. This is maybe a gap in my understanding of the way core and [+] libraries are paired.

**Comment by devngc** [comment]
noaacompare

**Comment by mostaphaRoudsari** [comment]
@devngc it all looks good. I'm going to merge this now and move your test cases to test folder later. Once we have that we can set-up call and discuss some of the details here and there. Great job overall!

**Comment by devngc** [comment]
@mostaphaRoudsari,
Thank you! Sure, it would be my pleasure to speak with you.

**Pull request merged**